### PR TITLE
Unwrap tool subtype in crafting tools

### PR DESCRIPTION
**Pull request by Coolthulhu**
First commit from #18882

Alone, this will cause a performance regression, so the binning PR should be merged as soon as possible after this one.

### CODE REVIEW

File: src/requirements.cpp [hunk 1]
```diff
@@ -333,6 +334,29 @@ void requirement_data::check_consistency()
     }
 }
 
+template<typename T>
+void requirement_data::finalize( std::vector< std::vector<T> > &vec )
```
Comment (mugling) [inline src/requirements.cpp]: YAGNI, certainly at least now now. It's likely large chunks of the crafting code will shortly be ripped out to support `AND`/`OR`

**Comment by Coolthulhu** [comment]
The failed check is that mod_manager.cpp astyle regression, not caused here.

File: src/item_factory.cpp [hunk 2]
```diff
@@ -263,6 +263,12 @@ void Item_factory::finalize() {
             }
         }
 
+        if( obj.tool != nullptr ) {
```
Comment (mugling) [inline src/item_factory.cpp]: Inconsistent with surround code. Should be `if( obj.tool )`

File: src/requirements.cpp [hunk 3]
```diff
@@ -333,6 +334,28 @@ void requirement_data::check_consistency()
...
+    for( auto &list : vec ) {
+        std::vector<tool_comp> new_list;
+        for( auto &comp : list ) {
+            const auto replacements = item_controller->subtype_replacement( comp.type );
+            for( const auto &replaced_type : replacements ) {
+                new_list.push_back( tool_comp( replaced_type, comp.count ) );
+            }
+        }
+
+        list = std::move( new_list );
```
Comment (mugling) [inline src/requirements.cpp]: I doubt `std::move` will do anything there - the compiler should be able to figure things out itself

File: src/requirements.cpp [hunk 4]
```diff
@@ -333,6 +334,28 @@ void requirement_data::check_consistency()
...
+    }
+}
+
+void requirement_data::finalize()
+{
+    for( auto &r : const_cast<std::map<requirement_id, requirement_data> &>( all() ) ) {
+        finalize( r.second.tools );
```
Comment (mugling) [inline src/requirements.cpp]: Could this not all be moved inline to avoid the extra function in the public API?

**Comment by mugling** [comment]
JSON load errors - bizarrely in `No_Acid_Zombies` mod?

**Comment by Coolthulhu** [comment]
Fixed and rebased.

**Comment by mugling** [comment]
I'll take another look

**Comment by mugling** [comment]
That fixed it

**Pull request merged**